### PR TITLE
Weekend-Codathon-29-08-2020

### PR DESCRIPTION
**Pull request by akanimax**

### CODE REVIEW

File: styleganv2/data/multi_image_loader.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    def __init__(
+        self,
+        data_dir: Path,
+        transform: Callable[[Image.Image], Tensor] = get_transform(),
+        input_data_range: Tuple[int, int] = (0, 1),
+        output_data_range: Tuple[int, int] = (-1, 1),
+        rec_dir: bool = False,
+    ) -> None:
+        # define the state of the object
+        self.rec_dir = rec_dir
```
Comment (akanimax) [inline styleganv2/data/multi_image_loader.py:81]: Oops buddy! Most of these need to be private... Was I drunk? :laughing:!

File: styleganv2/gan.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+                        and np.random.uniform() < mixing_regularization_prob
+                    ):
+                        g_loss = self._generator_mixing_regularization_loss(latents)
+                    else:
+                        fake = self.generator([latents for _ in range(self.depth)])
+                        g_loss = self.generator_loss(fake)
+
+                    gen_optim.zero_grad()
+                    g_loss.backward()
+                    if self._check_grad_ok(self.generator):
```
Comment (akanimax) [inline styleganv2/gan.py:284]: Make sure there isn't too much performance regression because of these checks

File: styleganv2/gan.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        self._toggle_all_networks("eval")
+
+        # create an EMA copy of the generator:
+        self.generator_shadow = copy.deepcopy(self.generator).to(self.device)
+
+        # initialize the gen_shadow weights equal to the
+        # weights of gen
+        update_average(self.generator_shadow, self.generator, beta=0)
+
+    def _toggle_all_networks(self, mode="train"):
```
Comment (akanimax) [inline styleganv2/gan.py:62]: good utility :smile:

File: styleganv2/gan.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+            only_inputs=True,
+            allow_unused=False,
+        )[0]
+
+        # calculate the penalty using these gradients
+        gradient = gradient.view(gradient.shape[0], -1)
+        gradient_penalty = torch.sum(gradient ** 2, dim=-1)
+        return gradient_penalty
+
+    def generator_loss(self, fake: Tensor) -> Tensor:
```
Comment (akanimax) [inline styleganv2/gan.py:108]: Loss functions are not customizable here :disappointed:! 
Perhaps a feature for the next codathon?

File: styleganv2/gan.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+                self.device
+            )
+        )
+        for replace_index in range(random_mix_point, len(latents_list)):
+            latents_list[replace_index] = second_latents
+        mix_latents_fake_samples = self.generator(latents_list)
+        return self.generator_loss(mix_latents_fake_samples)
+
+    def get_save_info(self):
+        if self.device == torch.device("cpu"):
```
Comment (akanimax) [inline styleganv2/gan.py:135]: This is ugly! :scream:. Rethink this

File: styleganv2/gan.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+                if not param_ok:
+                    grad_ok = False
+                    break
+        return grad_ok
+
+    @staticmethod
+    def normalize(x: Tensor, epsilon: float = 1e-8) -> Tensor:
+        return x / torch.sqrt(torch.mean(x ** 2, dim=-1, keepdim=True) + epsilon)
+
+    def train(
```
Comment (akanimax) [inline styleganv2/gan.py:169]: Wow.... this method is too long. Consider refactoring.

File: styleganv2/networks/discriminator.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        fmap_base: int = 16 << 10,
+        fmap_decay: float = 1.0,
+        fmap_min: int = 1,
+        fmap_max: int = 512,
+        use_minbatch_stddev: bool = False,
+    ) -> None:
+        super().__init__()
+
+        # object state
+        self.use_minbatch_stddev = use_minbatch_stddev
```
Comment (akanimax) [inline styleganv2/networks/discriminator.py:32]: Again check what can be private here ...

File: styleganv2/networks/generator.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        latent_size: int = 512,
+        fmap_base: int = 16 << 10,
+        fmap_decay: float = 1.0,
+        fmap_min: int = 1,
+        fmap_max: int = 512,
+    ) -> None:
+        super().__init__()
+
+        # object state
+        self.depth = depth
```
Comment (akanimax) [inline styleganv2/networks/generator.py:26]: same here

File: styleganv2/networks/shared/custom_layers.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            num_new_features:
+        """
+        super(MinibatchStdDev, self).__init__()
+        self.group_size = group_size
+        self.num_new_features = num_new_features
+
+    def extra_repr(self) -> str:
+        return f"group_size={self.group_size}, num_new_features={self.num_new_features}"
+
+    def forward(self, x: Tensor, alpha: float = 1e-8) -> Tensor:
```
Comment (akanimax) [inline styleganv2/networks/shared/custom_layers.py:152]: OMG wow! This is the slowest implementation of MinBatchStdDev I have ever seen. It's going to take magnitudes of more time compared to original TF implementation. Refer to the original code to make this better.

File: styleganv2/networks/shared/utils.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+from torch.nn import Module
+
+
+def nf(
```
Comment (akanimax) [inline styleganv2/networks/shared/utils.py:5]: Wow this is cool :sunglasses:!

**Review by akanimax — COMMENTED** [review]
Well, there are bunch of comments to take care of. Let's merge this PR into master for now, but Make sure all these comments are addressed in the next PR perhaps?
